File: nodejs-example/elastic_ingest_continuous.js
"use strict";
// Level 2: continuous Promise ingestion. Sample rows are regenerable; SDK memory is not durable source storage.
process.env.SS_LOG_LEVEL ??= "warn";
const { randomUUID } = require("node:crypto");
const streaming = require("snowpipe-streaming");
const RUN_ID = process.env.SNOWFLAKE_RUN_ID || randomUUID();

function connectionProperties() {
  const pat = process.env.SNOWFLAKE_PAT;
  if (!pat) return null;
  if (!process.env.SNOWFLAKE_ACCOUNT || !process.env.SNOWFLAKE_URL) {
    throw new Error("PAT authentication requires SNOWFLAKE_ACCOUNT and SNOWFLAKE_URL");
  }

  const properties = {
    authorization_type: "PAT",
    personal_access_token: pat,
    account: process.env.SNOWFLAKE_ACCOUNT,
    url: process.env.SNOWFLAKE_URL,
  };
  if (process.env.SNOWFLAKE_ROLE) {
    properties.role = process.env.SNOWFLAKE_ROLE;
  }
  return properties;
}

async function createClient() {
  const properties = connectionProperties();
  return streaming.createTableClient({
    clientName: `ingest-${randomUUID()}`,
    dbName: process.env.SNOWFLAKE_DATABASE || "MY_DATABASE",
    schemaName: process.env.SNOWFLAKE_SCHEMA || "MY_SCHEMA",
    tableName: process.env.SNOWFLAKE_TABLE || "MY_TABLE",
    ...(properties ? { properties } : { profilePath: process.env.SNOWFLAKE_PROFILE || "profile.json" }),
  });
}

function sampleRow(eventId) {
  return {
    EVENT_ID: eventId,
    C1: eventId,
    C2: `${RUN_ID}-${eventId}`,
  };
}

const MAX_PENDING = 1000;
const STALL_MS = 30 * 60_000;
const INVALID = new Set(["InvalidChannelError", "InvalidClientError", "ClosedClientError", "ClosedElasticChannelError"]);
const pause = () => new Promise((resolve) => setTimeout(resolve, 250));

// Observe outcomes immediately; never leave rejected Promises unhandled.
function submit(channel, eventId) {
  const item = { done: false, error: null };
  channel.appendRowWithWait(sampleRow(eventId), null).then(
    () => { item.done = true; },
    (error) => { item.error = error; item.done = true; },
  );
  return item;
}

// One control loop owns intake, progress, and recovery; SDK callbacks never do I/O.
async function main(clientFactory = createClient) {
  const total = Number(process.env.SNOWFLAKE_TEST_ROWS || 5000);
  if (!Number.isSafeInteger(total) || total < 0) throw new Error("Invalid row count");
  let stopping = false;
  const stop = () => { stopping = true; };
  process.on("SIGINT", stop);
  process.on("SIGTERM", stop);
  let client;
  let generation = 0;
  let attempts = 0;
  let nextId = 0;
  let confirmed = 0;
  let complete = false;
  let waitingForCapacity = false;
  let deadline = performance.now() + STALL_MS;
  const pending = new Map();

  try {
    client = await clientFactory();
    let channel = await client.getElasticChannel();

    while ((!stopping && nextId < total) || pending.size) {
      try {

        let progress = false;
        for (const [eventId, item] of pending) {
          if (!item.done) continue;
          if (item.error) throw item.error;
          pending.delete(eventId);
          confirmed++;
          progress = true;
        }
        if (progress || (!pending.size && !waitingForCapacity)) deadline = performance.now() + STALL_MS;
        if (performance.now() >= deadline) throw new Error("No durable progress for 30 minutes; retain unresolved events");
        if (!stopping && nextId < total && pending.size < MAX_PENDING) {
          // Replace sampleRow with a retained source read and your target-column mapping.
          pending.set(nextId, submit(channel, nextId));
          nextId++;
          waitingForCapacity = false;
          if (nextId % 100 === 0) await new Promise((resolve) => setImmediate(resolve));
        } else {
          await pause();
        }
      } catch (error) {
        if (error.httpStatusCode === 429) {
          waitingForCapacity = true;
          // A settled rejected append can be retried; unresolved appends must not be replayed.
          for (const [eventId, item] of pending) {
            if (item.done && item.error === error) {
              try { pending.set(eventId, submit(channel, eventId)); }
              catch (retry) { if (retry.httpStatusCode !== 429) throw retry; }
            }
          }
          if (performance.now() >= deadline) throw new Error("Backpressure persisted for 30 minutes");
          await pause();
          continue;
        }
        if (!INVALID.has(error.errorCode) || attempts++ >= 6) throw error;
        for (const [eventId, item] of pending) {
          if (item.done && !item.error) { pending.delete(eventId); confirmed++; }
        }
        try { await client.close({ waitForFlush: false, timeoutMs: 30_000 }); }
        catch (closing) { if (!INVALID.has(closing.errorCode)) throw closing; }
        client = null;
        generation++;
        client = await clientFactory();
        channel = await client.getElasticChannel();

        console.warn(`Recreated client; replaying ${pending.size} unresolved events; duplicates possible`);
        for (const eventId of pending.keys()) {
          while (true) {
            if (performance.now() >= deadline) throw new Error("Recovery exceeded stalled-progress budget");
            try { pending.set(eventId, submit(channel, eventId)); break; }
            catch (retry) { if (retry.httpStatusCode !== 429) throw retry; await pause(); }
          }
        }
      }
    }
    complete = true;
    console.log(`Durably acknowledged ${confirmed} rows; submitted=${nextId}; run=${RUN_ID}; stopped=${stopping}`);
  } finally {
    process.removeListener("SIGINT", stop);
    process.removeListener("SIGTERM", stop);
    if (!complete) console.error(`Retain source for replay; confirmed=${confirmed}, submitted=${nextId}`);
    if (client) await client.close({ waitForFlush: complete, timeoutMs: 30_000 });
  }
}

if (require.main === module) {
  const keepAlive = setInterval(() => {}, 1000);
  main().catch((error) => { console.error(error); process.exitCode = 1; })
    .finally(() => clearInterval(keepAlive));
}
module.exports = { main, createClient, sampleRow };

File: python-example/elastic_ingest_continuous.py
"""Level 2: continuous Futures with in-process recovery. SDK batching is automatic; source retention remains your responsibility."""

import os
import time
import uuid
from typing import Optional

os.environ.setdefault("SS_LOG_LEVEL", "warn")
from snowflake.ingest import streaming

DATABASE = os.environ.get("SNOWFLAKE_DATABASE", "MY_DATABASE")
SCHEMA = os.environ.get("SNOWFLAKE_SCHEMA", "MY_SCHEMA")
TABLE = os.environ.get("SNOWFLAKE_TABLE", "MY_TABLE")
PROFILE = os.environ.get("SNOWFLAKE_PROFILE", "profile.json")
RUN_ID = os.environ.get("SNOWFLAKE_RUN_ID", str(uuid.uuid4()))
Row = dict[str, object]

def connection_properties() -> Optional[dict[str, str]]:
    pat = os.environ.get("SNOWFLAKE_PAT")
    if not pat:
        return None
    if not os.environ.get("SNOWFLAKE_ACCOUNT") or not os.environ.get("SNOWFLAKE_URL"):
        raise ValueError("PAT authentication requires SNOWFLAKE_ACCOUNT and SNOWFLAKE_URL")

    properties = {
        "authorization_type": "PAT",
        "personal_access_token": pat,
        "account": os.environ["SNOWFLAKE_ACCOUNT"],
        "url": os.environ["SNOWFLAKE_URL"],
    }
    if os.environ.get("SNOWFLAKE_ROLE"):
        properties["role"] = os.environ["SNOWFLAKE_ROLE"]
    return properties


def create_client() -> streaming.StreamingIngestClient:
    properties = connection_properties()
    return streaming.StreamingIngestClient.from_table(
        client_name=f"ingest-{uuid.uuid4()}",
        db_name=DATABASE,
        schema_name=SCHEMA,
        table_name=TABLE,
        profile_json=None if properties else PROFILE,
        properties=properties,
    )


def sample_row(event_id: int) -> Row:
    return {
        "EVENT_ID": event_id,
        "C1": event_id,
        "C2": f"{RUN_ID}-{event_id}",
    }


MAX_PENDING = 1_000
STALL_SECONDS = 30 * 60
MAX_RECOVERIES = 6


def main():
    """Stream generated rows; pause on capacity, drain on interrupt, recover in process."""
    import signal

    total = int(os.environ.get("SNOWFLAKE_TEST_ROWS", "5000"))
    if total < 0:
        raise ValueError("SNOWFLAKE_TEST_ROWS must be nonnegative")
    stopping = False
    def stop_intake(signum, frame):
        nonlocal stopping
        stopping = True
    previous_signals = {sig: signal.signal(sig, stop_intake) for sig in (signal.SIGINT, signal.SIGTERM)}
    client = None
    pending = {}
    next_id = 0
    confirmed = 0
    attempts = 0
    generation = 0
    deadline = time.monotonic() + STALL_SECONDS
    complete = False
    waiting_for_capacity = False

    try:
        client = create_client()
        channel = client.get_elastic_channel()

        while next_id < total and not stopping or pending:
            try:

                # Observe outcomes without waiting for an entire submission window.
                progress = False
                for event_id, acknowledgement in list(pending.items()):
                    if not acknowledgement.done():
                        continue
                    acknowledgement.result()
                    del pending[event_id]
                    confirmed += 1
                    progress = True
                if progress or (not pending and not waiting_for_capacity):
                    deadline = time.monotonic() + STALL_SECONDS
                if time.monotonic() >= deadline:
                    raise TimeoutError("No durable progress for 30 minutes; retain unresolved source events")
                if next_id < total and not stopping and len(pending) < MAX_PENDING:
                    # Replace sample_row with a retained source read and row mapping.
                    event_id = next_id
                    pending[event_id] = channel.append_row_with_wait(sample_row(event_id), None)
                    next_id += 1
                    waiting_for_capacity = False
                else:
                    time.sleep(0.01)
            except streaming.StreamingIngestError as error:
                if error.http_status_code == 429:
                    # A terminally rejected append may be retried; untouched Futures stay pending.
                    for event_id, acknowledgement in list(pending.items()):
                        if acknowledgement.done() and acknowledgement.exception() is error:
                            try:
                                pending[event_id] = channel.append_row_with_wait(sample_row(event_id), None)
                            except streaming.StreamingIngestError as retry:
                                if retry.http_status_code != 429:
                                    raise
                    waiting_for_capacity = True
                    # Rejected append retains next_id; accepted work keeps its original outcome.
                    if time.monotonic() >= deadline:
                        raise TimeoutError("Backpressure persisted for 30 minutes") from error
                    time.sleep(0.25)
                    continue
                invalid = error.error_code.value in {
                    "InvalidChannelError", "InvalidClientError", "ClosedClientError", "ClosedElasticChannelError"
                }
                if not invalid or attempts >= MAX_RECOVERIES:
                    raise
                attempts += 1
                # First keep successes already observed; unknown outcomes may duplicate on replay.
                for event_id, acknowledgement in list(pending.items()):
                    if acknowledgement.done() and acknowledgement.exception() is None:
                        del pending[event_id]
                        confirmed += 1
                try:
                    client.close(wait_for_flush=False, timeout_seconds=30)
                except streaming.StreamingIngestError as closing:
                    if closing.error_code.value not in {"InvalidClientError", "ClosedClientError", "ClosedElasticChannelError"}:
                        raise
                client = None
                generation += 1
                client = create_client()
                channel = client.get_elastic_channel()

                # Regenerate this sample's unresolved rows. A real source must retain them.
                replay = list(pending)
                pending.clear()
                print(f"Recreated client; replaying {len(replay)} unresolved rows; duplicates possible")
                for event_id in replay:
                    while True:
                        if time.monotonic() >= deadline:
                            raise TimeoutError("Recovery exceeded stalled-progress budget")
                        try:
                            pending[event_id] = channel.append_row_with_wait(sample_row(event_id), None)
                            break
                        except streaming.StreamingIngestError as retry:
                            if retry.http_status_code != 429:
                                raise
                            time.sleep(0.25)
        complete = True
        print(f"Durably acknowledged {confirmed} rows; submitted={next_id}; run={RUN_ID}; stopped={stopping}")
    finally:
        for sig, handler in previous_signals.items():
            signal.signal(sig, handler)
        if not complete:
            print(f"Stopped with unconfirmed work; confirmed={confirmed}, submitted={next_id}; retain source for replay")
        if client is not None:
            client.close(wait_for_flush=complete, timeout_seconds=30)


if __name__ == "__main__":
    main()
